winch/codegen/src/codegen/control.rs

```diff
@@ -0,0 +1,154 @@
+//! results in the case of an empty then or else block:
+//!   (module
+//!     (func (export "params") (param i32) (result i32)
+//!       (i32.const 2)
+//!       (if (param i32) (result i32) (local.get 0)
+//!       (then))
+//!     (i32.const 3)
+//!     (i32.add)
+//!   )
+//!
+//! As a future optimization, we could perform a look ahead to the next
+//! instruction when reaching any of the comparison instructions. If the next
+//! instruction is a control instruction, we could avoid emitting
+//! a [`crate::masm::MacroAssembler::cmp_with_set`] and instead emit
+//! a conditional jump inline when emitting the control flow instruction.
+
+use super::{CodeGenContext, MacroAssembler, OperandSize};
+use crate::{
+    abi::{ABIResult, ABI},
+    masm::CmpKind,
+    CallingConvention,
+};
+use cranelift_codegen::MachLabel;
+use wasmtime_environ::WasmType;
+
+/// Holds the necessary metdata to support the emission
+/// of control flow instructions.
+pub(crate) enum ControlStackFrame {
+    If {
+        /// The continuation label.
+        ///
+        /// This could be either the else branch if
+        /// the if instruction has an else or the end
+        /// label otherwise.
+        cont: MachLabel,
+        /// The exit label, only present if there's an else
+        /// branch.
+        end: Option<MachLabel>,
+        /// The return values of the block.
+        result: ABIResult,
+        /// The size of the value stack at the beginning of the If.
+        original_stack_size: usize,
+    },
+}
+
+impl ControlStackFrame {
+    /// Returns [`ControlStackFrame`] for an if.
+    pub fn if_<M: MacroAssembler>(
+        returns: &[WasmType],
+        masm: &mut M,
+        context: &mut CodeGenContext,
+    ) -> Self {
+        let result = <M::ABI as ABI>::result(&returns, &CallingConvention::Default);
+        let mut control = Self::If {
+            cont: masm.get_label(),
+            end: None,
+            result,
+            original_stack_size: 0,
+        };
+
+        control.emit(masm, context);
+        control
+    }
+
+    fn emit<M: MacroAssembler>(&mut self, masm: &mut M, context: &mut CodeGenContext) {
+        match self {
+            ControlStackFrame::If {
+                cont,
+                original_stack_size,
+                ..
+            } => {
+                // Pop the condition value.
+                let top = context.pop_to_reg(masm, None, OperandSize::S32);
+
+                // Unconditionall spill before emitting control flow.
+                context.spill(masm);
+
+                *original_stack_size = context.stack.len();
+                masm.branch(CmpKind::Eq, top.into(), top.into(), *cont, OperandSize::S32);
+                context.free_gpr(top);
+            }
+        }
+    }
+
+    /// Handles the else branch if the current control stack frame is
+    /// [`ControlStackFrame::If`].
+    pub fn emit_else<M: MacroAssembler>(&mut self, masm: &mut M, context: &mut CodeGenContext) {
```

Comment:
I kind of like what `cranelift-wasm` does: it has another frame type `Else`, and the `else` opcode pops the `If` and pushes an `Else`. This keeps a 1-to-1 correspondence between the control-flow stack and the position in the code (i.e., the full context is captured) which is a little less error-prone I think. In this design with the methods on the frame struct itself, the `emit_else` method could mutate `self` (`*self = ControlStackFrame::Else { ... }`). This also removes the need for the conditional in the `end` case below and the somewhat subtle invariant around the option-of-label. What do you think?

Reply:
The resulting code is definitely cleaner with this approach, thanks!